src/libclient/deviceinfo_unix.cpp

```diff
@@ -56,78 +57,105 @@ DeviceInfo::~DeviceInfo()
-            break;
-        }
-        else
+        for (const char **fsentry = FSEntries; *fsentry != NULL; ++fsentry)
         {
-            // No UUID in fstab
-            QString target;
-            QFileInfo info;
+            fstab *tab = getfsfile(*fsentry);
 
-            if (uuid.indexOf("LABEL=") == 0)
+            if (!tab)
             {
-                uuid.remove(0, 6);
-                info = QFileInfo(QString("/dev/disk/by-label/%1").arg(QString::fromUtf8(uuid)));
-            }
-            else
-            {
-                info = QFileInfo(uuid);
+                continue;
             }
 
-            uuid.clear();
+            uuid = QByteArray::fromRawData(tab->fs_spec, strlen(tab->fs_spec));
 
-            if (info.isSymLink())
+            if (uuid.indexOf("UUID=") == 0)
             {
-                target = info.symLinkTarget();
+                uuid.remove(0, 5);
+                break;
             }
             else
             {
-                target = info.absoluteFilePath();
+                // No UUID in fstab
+                QString target;
+                QFileInfo info;
+
+                if (uuid.indexOf("LABEL=") == 0)
+                {
+                    uuid.remove(0, 6);
+                    info = QFileInfo(QString("/dev/disk/by-label/%1").arg(QString::fromUtf8(uuid)));
+                }
+                else
+                {
+                    info = QFileInfo(uuid);
+                }
+
+                uuid.clear();
+
+                if (info.isSymLink())
+                {
+                    target = info.symLinkTarget();
+                }
+                else
+                {
+                    target = info.absoluteFilePath();
+                }
+
+                uuid = uuidForDevice(target);
+
+                if (!uuid.isEmpty())
+                {
+                    break;
+                }
             }
+        }
 
-            uuid = uuidForDevice(target);
+        endfsent();
 
-            if (!uuid.isEmpty())
-            {
-                break;
-            }
+        if (uuid.isEmpty())
+        {
+            LOG_DEBUG("No HDD UID found!");
         }
+        else
+        {
+            hash.addData(uuid);
+            foundUUID = true;
+        }
+
     }
 
-    endfsent();
+    QList<QNetworkInterface> interfaces = QNetworkInterface::allInterfaces();
 
-    if (uuid.isEmpty())
+
+    foreach (const QNetworkInterface &inf, interfaces)
     {
-        LOG_DEBUG("No HDD UID found!");
-        return QString();
+        QString name = inf.name();
+        if (name.startsWith("eth") || name.startsWith("wlan")
+           || name.startsWith("wifi") || name.startsWith("en")
+           || name.startsWith("wl"))
+        {
+            hash.addData(inf.hardwareAddress().toUtf8());
+            foundMac = true;
+        }
     }
 
-    QCryptographicHash hash(QCryptographicHash::Sha224);
-    hash.addData(uuid);
+    if (foundUUID && !foundMac)
```

Comment:
* UUID && MAC -> hash (ok)
* UUID && !MAC -> empty string (seems wrong)
* !UUID && MAC -> hash (ok)
* !UUID && !MAC -> hash (definitely wrong)

I think it should be ```!foundUUID```.